File: CSharp/ExtensibleCommands/ExtensibleCommandsUnitTests/GenericExtensibleCommandsTest.cs
using System;
using System.Collections.Generic;
using Microsoft.VisualStudio.TestTools.UnitTesting;

using ExtensibleCommands;

namespace ExtensibleCommandsUnitTest
{
    [TestClass]
    public class GenericExtensibleCommandsTest
    {
        private TestContext testContextInstance;

        /// <summary>
        ///Gets or sets the test context which provides
        ///information about and functionality for the current test run.
        ///</summary>
        public TestContext TestContext
        {
            get
            {
                return testContextInstance;
            }
            set
            {
                testContextInstance = value;
            }
        }

        #region Additional test attributes
        // 
        //You can use the following additional attributes as you write your tests:
        //
        //Use ClassInitialize to run code before running the first test in the class
        [ClassInitialize()]
        public static void MyClassInitialize(TestContext testContext)
        {
            Setup.InitLog();
        }
        //
        //Use ClassCleanup to run code after all tests in a class have run
        //[ClassCleanup()]
        //public static void MyClassCleanup()
        //{
        //}
        //
        //Use TestInitialize to run code before running each test
        [TestInitialize()]
        public void MyTestInitialize()
        {
            Logger.Log(Logger.LogLevel.Info,
                "----------------------------------------------------------------------------------------------------------");
            Logger.Log(Logger.LogLevel.Info,
                string.Format("Starting Test {0}:{1}", GetType().Name, testContextInstance.TestName));
        }
        //
        //Use TestCleanup to run code after each test has run
        //[TestCleanup()]
        //public void MyTestCleanup()
        //{
        //}
        //
        #endregion

        // ------------------------------------------------------------------------------------------------
        // Helper classes

        private class Coord
        {
            public double X { get; set; }
            public double Y { get; set; }
            public double Z { get; set; }

            public Coord() { }

            public Coord(double x, double y, double z)
            {
                X = x;
                Y = y;
                Z = z;
            }
        }

        private class VisionSearchInput
        {
            public Coord TargetPosition { get; set; }
            public string PatternID { get; set; }

            public VisionSearchInput(Coord position, string patternID)
            {
                TargetPosition = position;
                PatternID = patternID;
            }
        }

        private class MotionController
        {
            /// <summary>
            /// Simulate Abort move
            /// </summary>
            public void Abort() { }

            /// <summary>
            /// Simulate Stage homing
            /// </summary>
            /// <param name="axis"></param>
            public void Home(string axis) { }

            /// <summary>
            /// Simulate Stage move
            /// </summary>
            /// <param name="axis"></param>
            /// <param name="position"></param>
            /// <returns></returns>
            public double Move(string axis, double position) { return position + 0.001; }
        }

        private class VisionProcessor
        {
            public double Search(string patternID)
            {
                double score = 95.3;
                return score;
            }
        }

        /// <summary>
        /// Simulation class for Motion System
        /// </summary>
        private class MotionSystem
        {
            public SimpleCommandIO<double, double> MoveXAxisCommand { get; private set; }
            public SimpleCommandIO<double, double> MoveYAxisCommand { get; private set; }
            public SimpleCommandIO<double, double> MoveZAxisCommand { get; private set; }

            public MotionSystem(MotionController motionController)
            {
                MoveXAxisCommand = new SimpleCommandIO<double, double>(position => motionController.Move("X", position), "Move X");
                MoveYAxisCommand = new SimpleCommandIO<double, double>(position => motionController.Move("Y", position), "Move Y");
                MoveZAxisCommand = new SimpleCommandIO<double, double>(position => motionController.Move("Z", position), "Move Z");
            }
        }

        /// <summary>
        /// Simulation class for Vision System
        /// </summary>
        private class VisionSystem
        {
            public SimpleCommandIO<VisionSearchInput, double> VisionSearchCommand { get; private set; }

            public VisionSystem(VisionProcessor visionProcessor)
            {
                VisionSearchCommand = new SimpleCommandIO<VisionSearchInput, double>(input => visionProcessor.Search(input.PatternID), "OutScore");
            }
        }

        /// <summary>
        /// Container for Pattern Search command
        /// </summary>
        private class PatternSearchCoordinator
        {
            public PatternSearchCommand PatternSearchCommand { get; private set; }

            public PatternSearchCoordinator(MotionSystem motionSystem, VisionSystem visionSystem)
            {
                PatternSearchCommand = new PatternSearchCommand(motionSystem, visionSystem);
            }
        }

        /// <summary>
        /// Custom command class implementing Pattern Search sequence
        /// </summary>
        private class PatternSearchCommand : SequentialCommand
        {
            // Input
            public Coord TargetPosition { set; get; }

            // Output
            public double VisionSearchScore { private set; get; }
            public Coord ActualPosition { private set; get; }

            public PatternSearchCommand(MotionSystem motionSystem, VisionSystem visionSystem)
            {
                Name = "Pattern Search";

                Add(new SimpleCommand(() => motionSystem.MoveXAxisCommand.Input = TargetPosition.X))    // Set input for X Move
                .Add(new SimpleCommand(() => motionSystem.MoveYAxisCommand.Input = TargetPosition.Y))   // Set input for Y Move
                .Add(new SimpleCommand(() => motionSystem.MoveZAxisCommand.Input = TargetPosition.Z))   // Set input for Z Move
                .Add(new ParallelCommand("Stage move")
                    .Add(motionSystem.MoveXAxisCommand)
                    .Add(motionSystem.MoveYAxisCommand)
                    .Add(motionSystem.MoveZAxisCommand))           // Do stage move (parallel X, Y, Z)             
                .Add(new SimpleCommand(() =>
                        visionSystem.VisionSearchCommand.Input = new VisionSearchInput(
                            new Coord(motionSystem.MoveXAxisCommand.Output,
                                motionSystem.MoveYAxisCommand.Output,
                                motionSystem.MoveZAxisCommand.Output),
                                "PATTERN_ID")
                    ))                                              // Set input for Vision Search
                .Add(visionSystem.VisionSearchCommand)              // Do Vision Search
                .Add(new SimpleCommand(() => 
                    { 
                        VisionSearchScore = visionSystem.VisionSearchCommand.Output;
                        ActualPosition = new Coord(motionSystem.MoveXAxisCommand.Output,
                            motionSystem.MoveYAxisCommand.Output,
                            motionSystem.MoveZAxisCommand.Output);
                    }));                                            // Set overall command output
            }
        }

        [TestMethod]
        public void StageInitializationTest()
        {
            // Simulate HW resources
            var mc = new MotionController();

            // Define a command to initialize stage parameters
            var initParametersCommand = new SimpleCommand(() => { /* Add parameter initialization! */}, "Init parameters");

            // Define abortable X/Y/Z Home commands
            var homeXAxisCommand = new AbortableCommand(new SimpleCommand(() => mc.Home("X")), mc.Abort, "Home X");

            var homeYAxisCommand = new AbortableCommand(new SimpleCommand(() => mc.Home("Y")), mc.Abort, "Home Y");

            var homeZAxisCommand = new AbortableCommand(new SimpleCommand(() => mc.Home("Z")), mc.Abort, "Home Z");

            // Define X/Y/Z retry Home commands
            var homeXAxisRetryCommand = new RetryCommand(homeXAxisCommand, 3);
            var homeYAxisRetryCommand = new RetryCommand(homeYAxisCommand, 3);
            var homeZAxisRetryCommand = new RetryCommand(homeZAxisCommand, 3);

            // Define abortable X/Y/Z Initial move commands
            var initMoveXAxisCommand = new AbortableCommand(new SimpleCommand(() => mc.Move("X", 0.0)), 
                mc.Abort, "Init Move X");

            var initMoveYAxisCommand = new AbortableCommand(new SimpleCommand(() => mc.Move("Y", 0.0)),
            mc.Abort, "Init Move Y");

            var initMoveZAxisCommand = new AbortableCommand(new SimpleCommand(() => mc.Move("Z", 0.0)),
            mc.Abort, "Init Move Z");

            // Define an command to do a parallel home and intial move of all 3 axes
            var homeAndMoveCommand = new ParallelCommand("Parallel Home and Move")
                .Add(new SequentialCommand("Home and Move X")
                        .Add(homeXAxisRetryCommand)
                        .Add(initMoveXAxisCommand))
                .Add(new SequentialCommand("Home and Move Y")
                        .Add(homeYAxisRetryCommand)
                        .Add(initMoveYAxisCommand))
                .Add(new SequentialCommand("Home and Move Z")
                        .Add(homeZAxisRetryCommand)
                        .Add(initMoveZAxisCommand));

            // Define an command to log stage initialization record
            var logRecordCommand = new SimpleCommand(() => { /* Add logging! */}, "Log record");

            // Define a complete Stage Initialization command
            var stageInitializationCommand = new TryCatchFinallyCommand(new SequentialCommand("Initialize")
                .Add(initParametersCommand).Add(homeAndMoveCommand), logRecordCommand, "Stage Initialization");

            stageInitializationCommand.Run();
        }

        [TestMethod()]
        public void ParameterInjectionTest()
        {
            // Simulate HW resources
            var patternSearchCoordinator = new PatternSearchCoordinator(new MotionSystem(new MotionController()),
                new VisionSystem(new VisionProcessor()));

            // Create main alignment command
            var cmd = patternSearchCoordinator.PatternSearchCommand;
            cmd.TargetPosition = new Coord(100.0, 20.0, -30.0);
            cmd.Run();

            Assert.AreEqual(95.3, cmd.VisionSearchScore);
            Assert.AreEqual(100.001, cmd.ActualPosition.X);
            Assert.AreEqual(20.001, cmd.ActualPosition.Y);
            Assert.AreEqual(-29.999, cmd.ActualPosition.Z);

            cmd.TargetPosition = new Coord(-10.0, 23.0, -3.0);
            cmd.Run();

            Assert.AreEqual(95.3, cmd.VisionSearchScore);
            Assert.AreEqual(-9.999, cmd.ActualPosition.X);
            Assert.AreEqual(23.001, cmd.ActualPosition.Y);
            Assert.AreEqual(-2.999, cmd.ActualPosition.Z);
        }
    }
}

File: Java/ExtensibleCommands/src/test/java/org/extensiblecommands/GenericExtensibleCommandsTest.java
package org.extensiblecommands;

import org.junit.*;
import org.junit.rules.TestName;

public class GenericExtensibleCommandsTest {
    @Rule
    public TestName name = new TestName();

    @BeforeClass
    public static void setUpClass() {
        Setup.InitLog();
    }

    @Before
    public void setUpTest() {
        Logger.log(Logger.LogLevel.Info,
                "----------------------------------------------------------------------------------------------------------");
        Logger.log(Logger.LogLevel.Info,
                String.format("Starting Test %s:%s", this.getClass().getName(), name.getMethodName()));
    }

    private class Coord {
        public double X;
        public double Y;
        public double Z;

        public Coord(double x, double y, double z) {
            X = x;
            Y = y;
            Z = z;
        }
    }

    private class VisionSearchInput {
        public Coord TargetPosition;
        public String PatternID;

        public VisionSearchInput(Coord position, String patternID) {
            TargetPosition = position;
            PatternID = patternID;
        }
    }

    private class MotionController {
        public void home(String axis) { }
        public Double move(String axis, Double position) { return position + 0.001; }
        public void abort() { }
    }

    private class VisionProcessor {
        public Double search(String patternID) {
            Double score = 95.3;
            return score;
        }
    }

    /**
     * Simulation class for Motion System
     */
    private class MotionSystem {
        public SimpleCommandIO<Double, Double> getMoveXAxisCommand() { return moveXAxisCommand; }
        public SimpleCommandIO<Double, Double> getMoveYAxisCommand() { return moveYAxisCommand; }
        public SimpleCommandIO<Double, Double> getMoveZAxisCommand() { return moveZAxisCommand; }

        private final SimpleCommandIO<Double, Double> moveXAxisCommand;
        private final SimpleCommandIO<Double, Double> moveYAxisCommand;
        private final SimpleCommandIO<Double, Double> moveZAxisCommand;

        public MotionSystem(MotionController motionController) {
            moveXAxisCommand = new SimpleCommandIO<>(position -> motionController.move("X", position), "Move X");
            moveYAxisCommand = new SimpleCommandIO<>(position -> motionController.move("Y", position), "Move Y");
            moveZAxisCommand = new SimpleCommandIO<>(position -> motionController.move("Z", position), "Move Z");
        }
    }

    /**
     * Simulation class for Vision System
     */
    private class VisionSystem {
        public SimpleCommandIO<VisionSearchInput, Double> getVisionSearchCommand() { return visionSearchCommand; }

        private final SimpleCommandIO<VisionSearchInput, Double> visionSearchCommand;

        public VisionSystem(VisionProcessor visionProcessor) {
            visionSearchCommand = new SimpleCommandIO<VisionSearchInput, Double>(input -> visionProcessor.search(input.PatternID), "OutScore");
        }
    }

    /**
     * Container for Pattern Search command
     */
    private class PatternSearchCoordinator {
        public PatternSearchCommand getPatternSearchCommand() { return patternSearchCommand; }

        private final PatternSearchCommand patternSearchCommand;

        public PatternSearchCoordinator(MotionSystem motionSystem, VisionSystem visionSystem) {
            patternSearchCommand = new PatternSearchCommand(motionSystem, visionSystem);
        }
    }

    /**
     * Custom command class implementing Pattern Search sequence
     */
    private class PatternSearchCommand extends SequentialCommand {
        public void setTargetPosition(Coord position) { targetPosition = position; }
        public double getVisionSearchScore() { return visionSearchScore; }
        public Coord getActualPosition() { return actualPosition; }

        public PatternSearchCommand(MotionSystem motionSystem, VisionSystem visionSystem) {
            name = "Alignment";
            // Create the main alignment command
            add(new SimpleCommand(() -> motionSystem.getMoveXAxisCommand().setInput(this.targetPosition.X)))
            .add(new SimpleCommand(() -> motionSystem.getMoveYAxisCommand().setInput(this.targetPosition.Y)))
            .add(new SimpleCommand(() -> motionSystem.getMoveZAxisCommand().setInput(this.targetPosition.Z)))
            .add(new ParallelCommand("Stage move")
                .add(motionSystem.getMoveXAxisCommand())
                .add(motionSystem.getMoveYAxisCommand())
                .add(motionSystem.getMoveZAxisCommand()))                  // Do stage move
            .add(new SimpleCommand(() -> visionSystem.getVisionSearchCommand().setInput(
                new VisionSearchInput(
                    new Coord(motionSystem.getMoveXAxisCommand().getOutput(),
                        motionSystem.getMoveYAxisCommand().getOutput(),
                        motionSystem.getMoveZAxisCommand().getOutput()),
                        "PATTERN_ID")
                )))      // Do vision search
            .add(visionSystem.getVisionSearchCommand())
            .add(new SimpleCommand(() -> {
                visionSearchScore = visionSystem.getVisionSearchCommand().getOutput();
                actualPosition = new Coord(motionSystem.getMoveXAxisCommand().getOutput(),
                    motionSystem.getMoveYAxisCommand().getOutput(),
                    motionSystem.getMoveZAxisCommand().getOutput());
                    }));
        }

        private Coord targetPosition;
        private double visionSearchScore;
        private Coord actualPosition;
    }

    @Test
    public void stageInitializationTest() throws Exception {
        // Simulate HW resources
        var mc = new MotionController();

        // Define a command to initialize stage parameters
        var initParametersCommand = new SimpleCommand(() -> { /* Add parameter initialization! */}, "Init parameters");

        // Define abortable X/Y/Z home commands
        var homeXAxisCommand = new AbortableCommand(new SimpleCommand(() -> mc.home("X")),
                () -> mc.abort(), "home X");

        var homeYAxisCommand = new AbortableCommand(new SimpleCommand(() -> mc.home("Y")),
                () -> mc.abort(), "home Y");

        var homeZAxisCommand = new AbortableCommand(new SimpleCommand(() -> mc.home("Z")),
                () -> mc.abort(), "home Z");

        // Define X/Y/Z retry home commands
        var homeXAxisRetryCommand = new RetryCommand(homeXAxisCommand, 3);
        var homeYAxisRetryCommand = new RetryCommand(homeYAxisCommand, 3);
        var homeZAxisRetryCommand = new RetryCommand(homeZAxisCommand, 3);

        // Define abortable X/Y/Z Initial move commands
        var initMoveXAxisCommand = new AbortableCommand(new SimpleCommand(() -> mc.move("X", 0.0)),
        () -> mc.abort(), "Init move X");

        var initMoveYAxisCommand = new AbortableCommand(new SimpleCommand(() -> mc.move("Y", 0.0)),
        () -> mc.abort(), "Init move Y");

        var initMoveZAxisCommand = new AbortableCommand(new SimpleCommand(() -> mc.move("Z", 0.0)),
        () -> mc.abort(), "Init move Z");

        // Define an command to do a parallel home and intial move of all 3 axes
        var homeAndMoveCommand = new ParallelCommand("Parallel home and move")
                .add(new SequentialCommand("home and move X")
                        .add(homeXAxisRetryCommand)
                        .add(initMoveXAxisCommand))
                .add(new SequentialCommand("home and move Y")
                        .add(homeYAxisRetryCommand)
                        .add(initMoveYAxisCommand))
                .add(new SequentialCommand("home and move Z")
                        .add(homeZAxisRetryCommand)
                        .add(initMoveZAxisCommand));

        // Define an command to log stage initialization record
        var logRecordCommand = new SimpleCommand(() -> { /* Add logging! */}, "Log record");

        // Define a complete Stage Initialization command
        var stageInitializationCommand = new TryCatchFinallyCommand(new SequentialCommand("Initialize")
                .add(initParametersCommand).add(homeAndMoveCommand), logRecordCommand, "Stage Initialization");

        stageInitializationCommand.run();
    }

    @Test
    public void ParameterInjectionTest() throws Exception {
        // Simulate HW resources
        var patternSearchCoordinator = new PatternSearchCoordinator(new MotionSystem(new MotionController()),
                new VisionSystem(new VisionProcessor()));

        // Create main alignment command
        var cmd = patternSearchCoordinator.getPatternSearchCommand();
        cmd.setTargetPosition(new Coord(100.0, 20.0, -30.0));
        cmd.run();

        Assert.assertEquals(95.3, cmd.getVisionSearchScore(), 0.001);
        Assert.assertEquals(100.001, cmd.getActualPosition().X, 0.001);
        Assert.assertEquals(20.001, cmd.getActualPosition().Y, 0.001);
        Assert.assertEquals(-29.999, cmd.getActualPosition().Z, 0.001);

        cmd.setTargetPosition(new Coord(-10.0, 23.0, -3.0));
        cmd.run();

        Assert.assertEquals(95.3, cmd.getVisionSearchScore(), 0.001);
        Assert.assertEquals(-9.999, cmd.getActualPosition().X, 0.001);
        Assert.assertEquals(23.001, cmd.getActualPosition().Y, 0.001);
        Assert.assertEquals(-2.999, cmd.getActualPosition().Z, 0.001);
    }

    public void HelloWorldTest() throws Exception {
        // Output string to console
        var helloWordlCmd = new SimpleCommandI<String>(input -> System.out.println(input), "Hello World\n");

        // Wait for user input
        var waitForConsoleInputCmd = new SimpleCommand(() -> System.in.read());

        // Create sequence of the above 2 steps
        var sequentialCommand = new SequentialCommand();
        sequentialCommand.add(helloWordlCmd).add(waitForConsoleInputCmd);

        // Supply input and run sequence
        helloWordlCmd.setInput("Hello World!");
        sequentialCommand.run();
    }
}
